### ISSUE TITLE
Tailscale discovery not working

### ISSUE THREAD

**Issue by Mansarde** [charmbracelet/wishlist#217]
OS version: **Windows 11** (Build 22621)
Wishlist version: **0.13.0**

I have tried using wishlist in combination with Tailscale, but it doesn't seem to work.
### Here's what I did:
1. Create an OAuth key in the Tailscale admin console, let's call its id `<OAUTH_ID>` and its key `tskey-client-<OAUTH_ID>-<SECRET>`
![image](https://github.com/charmbracelet/wishlist/assets/847687/1f8ee9ef-7350-42f0-9303-e6678b5189df)
2. On a machine with Tailscale installed and connected, I typed this command:
`wishlist --tailscale.net=<REDACTED>.ts.net --tailscale.client.id=<OAUTH_ID> --tailscale.client.secret=tskey-client-<OAUTH_ID>-<SECRET>`

---
But nothing happens, I'm still on the command line as if nothing was executed.
Doing a `echo %errorlevel%` afterwards returns a `1`, which denotes that there was an error.
But the console does not output anything to give me a hint as to what went wrong.

The logs are not illuminating either, as it looks like there is no error:
```
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"<REDACTED>.ts.net","ts":"2023/08/15 17:57:24"}
{"lvl":"info","msg":"using oauth","ts":"2023/08/15 17:57:24"}
```
---
I also tried it with a temporary auth key instead (i.e. using `--tailscale.key` instead of `--tailscale.client.id` and `--tailscale.client.secret`), but that also doesn't do anything.
Other than that I tried to provide the arguments in different ways (e.g. with a space or `=`, with quotes and without), but nothing had any effect.

Only if I pass the example `config.yml` via `-c` will Wishlist actually start up.
But if I my Tailscale arguments together with the config file, Wishlist silently fails to start.

Am I doing anything wrong? 🤔

EDIT: I noticed that using an incorrect secret or id will display the exact same behavior, with no indication of an error (neither on CLI  or log file).

**Comment by caarlos0** [comment]
I cannot reproduce this here, tried multiple ways and every time it either worked or returned with an error...

can you paste the full logs?

**Comment by Mansarde** [comment]
Sure thing, here you go:
<details><summary>Click me</summary>

```
{"lvl":"info","msg":"Using configuration file","path":"C:\\Users\\HomePC\\AppData\\Roaming\\wishlist\\wishlist.yml","ts":"2023/08/15 17:23:21"}
{"height":4,"lvl":"debug","msg":"setting delegate height","ts":"2023/08/15 17:23:21"}
{"lvl":"info","msg":"Using configuration file","path":"E:\\Portables\\charm-apps\\wishlist\\config.yml","ts":"2023/08/15 17:23:41"}
{"height":4,"lvl":"debug","msg":"setting delegate height","ts":"2023/08/15 17:23:41"}
{"err":"failed to parse config: yaml: line 6: did not find expected key","lvl":"info","msg":"Not using","path":"E:\\Portables\\charm-apps\\wishlist\\config.yml","ts":"2023/08/15 17:24:16"}
{"lvl":"info","msg":"Using configuration file","path":"E:\\Portables\\charm-apps\\wishlist\\config.yml","ts":"2023/08/15 17:24:22"}
{"height":4,"lvl":"debug","msg":"setting delegate height","ts":"2023/08/15 17:24:22"}
{"lvl":"info","msg":"Using configuration file","path":"E:\\Portables\\charm-apps\\wishlist\\config.yml","ts":"2023/08/15 17:25:02"}
{"height":4,"lvl":"debug","msg":"setting delegate height","ts":"2023/08/15 17:25:02"}
{"lvl":"info","msg":"Using configuration file","path":"E:\\Portables\\charm-apps\\wishlist\\config.yml","ts":"2023/08/15 17:25:40"}
{"height":4,"lvl":"debug","msg":"setting delegate height","ts":"2023/08/15 17:25:40"}
{"lvl":"info","msg":"Using configuration file","path":"E:\\Portables\\charm-apps\\wishlist\\config.yml","ts":"2023/08/15 17:43:20"}
{"height":4,"lvl":"debug","msg":"setting delegate height","ts":"2023/08/15 17:43:20"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:46:07"}
{"lvl":"info","msg":"using oauth","ts":"2023/08/15 17:46:07"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:46:18"}
{"lvl":"info","msg":"using oauth","ts":"2023/08/15 17:46:18"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:48:30"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:49:46"}
{"lvl":"info","msg":"using api key auth","ts":"2023/08/15 17:49:46"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:50:25"}
{"lvl":"info","msg":"using api key auth","ts":"2023/08/15 17:50:25"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:50:27"}
{"lvl":"info","msg":"using api key auth","ts":"2023/08/15 17:50:27"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:50:36"}
{"lvl":"info","msg":"using api key auth","ts":"2023/08/15 17:50:36"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:51:12"}
{"lvl":"info","msg":"using api key auth","ts":"2023/08/15 17:51:12"}
{"lvl":"info","msg":"Using configuration file","path":"E:\\Portables\\charm-apps\\wishlist\\config.yml","ts":"2023/08/15 17:51:21"}
{"height":4,"lvl":"debug","msg":"setting delegate height","ts":"2023/08/15 17:51:21"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:51:39"}
{"lvl":"info","msg":"using api key auth","ts":"2023/08/15 17:51:39"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:52:48"}
{"lvl":"info","msg":"using api key auth","ts":"2023/08/15 17:52:48"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:54:55"}
{"lvl":"info","msg":"using oauth","ts":"2023/08/15 17:54:55"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:56:01"}
{"lvl":"info","msg":"using oauth","ts":"2023/08/15 17:56:01"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:56:37"}
{"lvl":"info","msg":"using api key auth","ts":"2023/08/15 17:56:37"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 17:57:24"}
{"lvl":"info","msg":"using oauth","ts":"2023/08/15 17:57:24"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 18:16:59"}
{"lvl":"info","msg":"using oauth","ts":"2023/08/15 18:16:59"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 18:17:20"}
{"lvl":"info","msg":"using oauth","ts":"2023/08/15 18:17:20"}
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"REDACTED.ts.net","ts":"2023/08/15 18:17:27"}
{"lvl":"info","msg":"using oauth","ts":"2023/08/15 18:17:27"}

```
</details>

I only replaced my actual Tailnet name with `REDACTED`, for privacy.
The config errors in the beginning are just because I tried to put the Tailscale values into it, but couldn't get it to work.
So I ended up scrapping the config file and just went with the command line arguments, as described in the OP.

**Comment by caarlos0** [comment]
the Tailscale options are only available in the command line / env vars... 

can you share the problematic configuration file as well? 

also, you're running in local mode, yes? can you also share the full commands you ran (both the ones that work and the ones that didn't)?

thank you for reporting this btw 💜

**Comment by Mansarde** [comment]
With [this config.yml](https://github.com/charmbracelet/wishlist/files/12353178/config.yml.txt), I successfully ran Wishlist by entering:
`wishlist -c "E:\charm-apps\wishlist\config.yml"`

It shows the dummy entry I specified in the config (just to see if it actually uses my config file):
![image](https://github.com/charmbracelet/wishlist/assets/847687/34af564f-1340-4f4b-9b1c-29a4cc3aa76d)

The log will also produce these two entries:
```
{"lvl":"info","msg":"Using configuration file","path":"E:\\charm-apps\\wishlist\\config.yml","ts":"2023/08/15 23:13:30"}
{"height":3,"lvl":"debug","msg":"setting delegate height","ts":"2023/08/15 23:13:30"}
```

If I try to add the tailscale arguments, like so:
`wishlist -c "E:\charm-apps\wishlist\config.yml" --tailscale.net=<MY_TAILNET>.ts.net --tailscale.client.id=<OAUTH_ID> --tailscale.client.secret=<OAUTH_KEY>`

Then nothing happens in the CLI and the logs produce:
```
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"<MY_TAILNET>.ts.net","ts":"2023/08/15 23:17:04"}
{"lvl":"info","msg":"using oauth","ts":"2023/08/15 23:17:04"}
```

If I try the Tailscale arguments _without_ providing a config file, I get the exact same log entries.
And the same thing also happens if I replace both `--tailscale.client.id` and `--tailscale.client.secret` with `--tailscale.key`, except that instead of `"using oauth"`, the log will say `"using api key auth"` (with everything else remaining the same.
I also tried it with the env vars `TAILSCALE_CLIENT_ID` and `TAILSCALE_CLIENT_SECRET`, but to no avail. :(

---
Would any settings in Tailscale have any effect on Wishlist?
I do have some custom DNS settings set up in the Tailscale admin console, like split DNS and a search domain.
Or do we have to adjust the ACLs in Tailscale?

What is weird to me is that when I tried Wishlist with `--tailscale.key` then I made sure to set it to "Pre-Approved".
But no device ever pops up in my device list in the Tailscale admin console, implying that the key has never been used by Wishlist. 🤔

I might spin up a VM and try the same in there to see if I get the same problem.
Maybe it's just a very coincidental fluke on my main desktop PC.
I'll report back as soon as I gave it a try.

**Comment by Mansarde** [comment]
Ok, I used a fresh and up-to-date Windows 10 VM, but had the exact same problem there.
I.e. I installed and logged into Tailscale, downloaded a fresh Wishlist, executed it with the Tailscale arguments, but got the same result as described before.

Next on my list of things to test is to try it on a completely different _device_ (e.g. a laptop) and see if it behaves the same.
If it does, then I'm going to try _another Tailscale account_. Currently I'm using a "Github Corporation" account, so just to make sure I'll check if the same problem persists with another account.
If that still won't work, then I have to try it on _another network_, e.g. next time I'll visit a friend's house.
<sub>(Aside from that I might try it on a Linux device, just to see if I can get it working at all, though ultimately I'd like to use it from a Windows machine.)</sub>

I'll try all of those and report back once I got some useful data.

**Comment by Mansarde** [comment]
Hey there, sorry for the late response, IRL got a bit in the way.
I've not gotten any luck with the other tests, always the same results, but....
Your last comment did the trick to finding out what the problem was and now everything works! 🥳

So first I tried your CURL command, but there was no output.
After removing the silent (`-s`) flag, the command responded with:
`{"message":"user tailnet does not match"}`
Confused, I triple checked my tailnet and I didn't mistype anything. I even copied it straight from the DNS page (in the admin console) just to make sure.

After looking into Tailscale's [API docs](https://github.com/tailscale/tailscale/blob/main/api.md#tailnet), I discovered that for API calls involving the _tailnet name_, you can either:
  - Provide a dash (`-`) to reference the default tailnet [...]
  - or provide the ***organization*** name [...]

Once I used either of these two , I got a successful response from the CURL command which showed me all my devices!
Right after I tried the organization name (and even the dash) in the Wishlist command and... whadda ya know, Wishlist finally started _and_ showed me all the devices on my tailnet! 😁

### tl;dr
I had to use the _organization_ name (in my case a `*.org.github` domain) instead of my actual _tailnet name_ (a `*.ts.net` domain).

---

So this can be chalked up to user error, my apologies.
Though I have to say it is a bit confusing that the Tailscale API docs say "tailnet" when they really mean "organization".
Especially since every other part of the (non-API) docs (as well as the DNS settings page) refer to the `*.ts.net` domain as one's tailnet.
Ah well, TIL. ¯\_(ツ)_/¯

Anyways, the hint with CURL'ing the Tailscale API was very helpful. That pointed me in the right direction and did the trick to find out where I was going wrong.
Thanks a lot! 🙂✌️

---
P.S.:
The logs are also now showing a line after devices were discovered (the third line):
```
{"lvl":"debug","msg":"discovering from tailscale","tailnet":"-","ts":"2023/08/30 00:51:51"}
{"lvl":"info","msg":"using oauth","ts":"2023/08/30 00:51:51"}
{"devices":12,"lvl":"info","msg":"discovered from tailscale","tailnet":"-","ts":"2023/08/30 00:51:51"}
```
It would be nice if Wishlist logged error responses from the Tailscale API (instead of silently failing).
Like e.g. the API response I mentioned above: `{"message":"user tailnet does not match"}`

**Comment by caarlos0** [comment]
ah got it

OK now I noticed why I couldn't repro, you're running in local mode, I was trying server mode.

seem like local mode is not logging that error for some reason, will take a look! Thanks!
